### PR TITLE
Account for DF limits per channel in DynamicFilterSourceOperator

### PR DESCRIPTION
**Pull request by skrzypo987**
## Description
Up to this point, when dynamic filter went beyond the threshold all channels fall back to the min/max range predicate. Now the biggest channel falls back, while the rest continue to be gathered as a set of values. This way if a single channel is too big, it won't prevent other, smaller ones to be used as a dynamic filter.

## Additional context and related issues
Fixes https://github.com/trinodb/trino/issues/14797

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# General
* Improve performance of queries with selective joins. ({issue}`15569`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java [hunk 1]
```diff
@@ -487,4 +487,106 @@ public boolean isFinished()
             return current == null && finished;
         }
     }
+
+    // Simple object to encapsulate whether channel is handled by set, range or neither
+    private class ChannelsState
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:492]: Can we instead make a per channel `ChannelFilter` class and make the processing of block and fallback logic an internal detail of this class ?

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java [hunk 2]
```diff
@@ -234,78 +227,83 @@ public void addInput(Page page)
...
+                if (channelsState.get(channelIndex) == ChannelState.SET) {
+                    filterSizeInBytes += valueSets[channelIndex].getRetainedSizeInBytes();
+                    filterMaxDistinctValues = Math.max(filterMaxDistinctValues, valueSets[channelIndex].size());
+                }
+            }
         }
     }
 
-    private void handleTooLargePredicate()
+    private int getBiggestChannel(boolean biggestRetainedSize)
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:279]: It seems overly complicated to implement fallback for biggest channel.
Can we instead make a per channel `ChannelFilter` class and let each channel individually fallback to min-max and then no collection based on the distinct values and row count seen by it ?

Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:279]: The problem with this approach is that the we need to either add more config options to an already bloated config class or change the meaning of the existing ones.
The solution with getting rid of the biggest df is the least problematic one

Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:279]: I would rather remove/rename/repurpose configs to get simpler code here than make this code even more complex than it already is.
Typically we expect users to use `enable-large-dynamic-filters`, the rest of the configs are a last resort to allow dynamic filtering for specific scenarios. I don't see those configs as as some immutable contract.

Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:279]: @sopel39 Please share your opinion here

**Comment by wendigo** [comment]
@skrzypo987 can you share some benchmarks?

**Comment by skrzypo987** [comment]
> @skrzypo987 can you share some benchmarks?

I haven't done any. Making a benchmark that will greatly benefit from this change is easy, but doesn't prove anything since its written exactly for this case.

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java (hunk 2 above)
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:279]: > The problem with this approach is that the we need to either add more config options to an already bloated config class or change the meaning of the existing ones.

I think we can keep current properties, but apply them per channel.
Per channel limits seem to make more sense.

Let's take semi join as an example:
plan1
```
semijoin(l1 = r1)
  semijoin(l2 = r2)
```
vs plan2
```
semijoin(l1 = r1 and l2 = r2)
```
it doesn't make sense that `plan2` has smaller limits than plan1

**Comment by sopel39** [comment]
> Making a benchmark that will greatly benefit from this change is easy, but doesn't prove anything since its written exactly for this case.

In this case we are also interested if the change doesn't introduce any regressions

**Comment by skrzypo987** [comment]
> > Making a benchmark that will greatly benefit from this change is easy, but doesn't prove anything since its written exactly for this case.
> 
> In this case we are also interested if the change doesn't introduce any regressions

If you mean TPC benchmarks, I can run them just to be sure. But I don't think any of it will trigger this behaviour

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java (hunk 2 above)
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:279]: All right, I will rewrite this to be on a per-channel basis and ping you when its done. The code should indeed be simpler.

**Comment by sopel39** [comment]
> If you mean TPC benchmarks, I can run them just to be sure. But I don't think any of it will trigger this behaviour

There are few queries with multi channel joins there.

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java [hunk 3]
```diff
@@ -487,4 +339,163 @@ public boolean isFinished()
             return current == null && finished;
         }
     }
+
+    private class ChannelFilter
+    {
+        private final int channelIndex;
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:345]: We don't need `channelIndex`, we can keep `type` as a member variable instead.

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java [hunk 4]
```diff
@@ -487,4 +339,163 @@ public boolean isFinished()
...
+
+        private ChannelState state;
+        private boolean minMaxPossible;
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:348]: minMaxPossible -> collectMinMax

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java [hunk 5]
```diff
@@ -370,42 +242,16 @@ public void finish()
...
-            Block block = blockBuilders[channelIndex].build();
-            Type type = channels.get(channelIndex).type;
-            domainsBuilder.put(channels.get(channelIndex).filterId, convertToDomain(type, block));
+            DynamicFilterId filterId = channels.get(channelIndex).filterId;
+            channelFilters[channelIndex].getDomain().ifPresent(domain -> domainsBuilder.put(filterId, domain));
         }
-        valueSets = null;
-        blockBuilders = null;
         dynamicPredicateConsumer.addPartition(TupleDomain.withColumnDomains(domainsBuilder.buildOrThrow()));
     }
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:257]: `convertToDomain` should now be a private static method or a private method of `ChannelFilter`

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java [hunk 6]
```diff
@@ -234,123 +206,23 @@ public void addInput(Page page)
...
-                BlockPositionComparison comparison = minMaxComparisons.get(i);
-                Block block = page.getBlock(channels.get(channelIndex).index);
-                updateMinMaxValues(block, i, comparison);
-            }
+
+        if (emptyDomain) {
             return;
         }
-        minMaxCollectionLimit -= page.getPositionCount();
-        // TODO: we should account for the memory used for collecting build-side values using MemoryContext
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:257]: TODO should be restored as it is still needed

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java [hunk 7]
```diff
@@ -234,123 +206,23 @@ public void addInput(Page page)
...
-        long filterSizeInBytes = 0;
-        int filterMaxDistinctValues = 0;
-        // Collect only the columns which are relevant for the JOIN.
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:260]: This comment is still useful

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java [hunk 8]
```diff
@@ -234,123 +206,23 @@ public void addInput(Page page)
...
-            if (comparison.compare(block, position, block, minValuePosition) < 0) {
-                minValuePosition = position;
-            }
-            else if (comparison.compare(block, position, block, maxValuePosition) > 0) {
-                maxValuePosition = position;
+        if (minMaxCollectionLimit < 0) {
+            for (int channelIndex = 0; channelIndex < channels.size(); channelIndex++) {
+                channelFilters[channelIndex].disableMinMax();
             }
         }
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:221]: This should move inside the `if (minMaxCollectionLimit >= 0)` block, it doesn't need to be called more than once

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java [hunk 9]
```diff
@@ -487,4 +339,163 @@ public boolean isFinished()
...
+        @Nullable
+        private TypedSet valueSet;
+        @Nullable
+        private Block minValues;
+        @Nullable
+        private Block maxValues;
+        @Nullable
+        private BlockPositionComparison minMaxComparison;
+
+        public ChannelFilter(Channel channel, boolean minMaxEnabled, PlanNodeId planNodeId, BlockTypeOperators blockTypeOperators, int channelIndex)
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:360]: We don't need to make any method in this class `public`

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java [hunk 10]
```diff
@@ -487,4 +339,163 @@ public boolean isFinished()
...
+            if (comparison.compare(block, minValuePosition, currentMin, 0) < 0) {
+                minValues = block.getSingleValueBlock(minValuePosition);
+            }
+            if (comparison.compare(block, maxValuePosition, currentMax, 0) > 0) {
+                maxValues = block.getSingleValueBlock(maxValuePosition);
+            }
+        }
+    }
+
+    private void checkIfEmpty()
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:486]: Move this below the `isFinished` method

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java [hunk 11]
```diff
@@ -487,4 +339,163 @@ public boolean isFinished()
...
+        }
+
+        public void disableMinMax()
+        {
+            minMaxPossible = false;
+            if (state == ChannelState.MIN_MAX) {
+                state = ChannelState.NONE;
+                minMaxChannels--;
+            }
+            checkIfEmpty();
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:434]: We don't need to call `checkIfEmpty` here, it can be done once after the for loop on `disableMinMax` in `addInput`

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java [hunk 12]
```diff
@@ -487,4 +339,163 @@ public boolean isFinished()
             return current == null && finished;
         }
     }
+
+    private class ChannelFilter
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:343]: It would be cleaner for this class to be `static`. When inner class can access anything from the outer one, the contract between them is unclear.
You can pass down a reference to `checkEmpty` via a `Runnable notifyCompletion` or something similar.

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java [hunk 13]
```diff
@@ -155,20 +156,12 @@ public OperatorFactory duplicate()
     private final long maxFilterSizeInBytes;
 
     private final List<Channel> channels;
-    private final List<Integer> minMaxChannels;
-    private final List<BlockPositionComparison> minMaxComparisons;
-
-    // May be dropped if the predicate becomes too large.
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:161]: This comment can move to ChannelFilter

File: core/trino-main/src/test/java/io/trino/operator/TestDynamicFilterSourceOperator.java [hunk 14]
```diff
@@ -440,7 +440,7 @@ public void testMultipleColumnsCollectBelowDistinctValuesLimit()
     }
 
     @Test
-    public void testMultipleColumnsCollectMinMaxRangeWhenTooManyDistinctValues()
+    public void testMultipleColumnsBothDfAndRangeWhenTooManyDistinctValues()
```
Comment (raunaqmorarka) [inline core/trino-main/src/test/java/io/trino/operator/TestDynamicFilterSourceOperator.java:443]: Df -> Set

File: core/trino-main/src/test/java/io/trino/operator/TestDynamicFilterSourceOperator.java [hunk 15]
```diff
@@ -452,11 +452,28 @@ public void testMultipleColumnsCollectMinMaxRangeWhenTooManyDistinctValues()
...
                         new DynamicFilterId("0"), Domain.create(ValueSet.ofRanges(
                                 range(BIGINT, 0L, true, 100L, true)), false),
+                        new DynamicFilterId("1"), Domain.singleValue(COLOR, 100L),
                         new DynamicFilterId("2"), Domain.create(ValueSet.ofRanges(
                                 equal(BIGINT, 200L)), false))));
         assertDynamicFilters(maxDistinctValues, ImmutableList.of(BIGINT, COLOR, BIGINT), ImmutableList.of(largePage), expectedTupleDomains);
     }
 
+    @Test
+    public void testMultipleColumnsSingleDFWithNoRangeWhenTooManyDistinctValues()
```
Comment (raunaqmorarka) [inline core/trino-main/src/test/java/io/trino/operator/TestDynamicFilterSourceOperator.java:462]: SingleDF -> Set

**Review by raunaqmorarka — COMMENTED** [review]
Please share results from JMH BenchmarkDynamicFilterSourceOperator as well (tweak limits if necessary there to actually exercise the fallback code paths).

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java (hunk 8 above)
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:221]: I don't follow, please be more specific about what is "it"

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java (hunk 9 above)
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:360]: I prefer to make methods that are used outside of this class public, for readability reasons.

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java (hunk 12 above)
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:343]: What is wrong with it being non-static. This is an implementation detail, inner class made to group code better. Passing some Runnables around, on the other hand, seems excessive and overly complicated IMHO.

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java [hunk 16]
```diff
@@ -487,4 +339,163 @@ public boolean isFinished()
...
+                    if (valueSet.getRetainedSizeInBytes() > maxFilterSizeInBytes || valueSet.size() > maxDistinctValues) {
+                        dfChannels--;
+                        if (minMaxPossible) {
+                            updateMinMaxValues(blockBuilder.build(), minMaxComparison);
+                            minMaxChannels++;
+                            state = ChannelState.MIN_MAX;
+                        }
+                        else {
+                            checkIfEmpty();
+                            state = ChannelState.NONE;
```
[the remark this answers is not in the document]
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:395]: nice catch

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java (hunk 10 above)
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:486]: It is used by `ChannelFilter` class which is below `isFinished`

**Review by skrzypo987 — COMMENTED** [review]
@raunaqmorarka AC

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java (hunk 8 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java]: "it" is the whole `if (minMaxCollectionLimit < 0)` block. It looks like we will call it for every page after `minMaxCollectionLimit < 0` condition is met even though that's not needed.

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java [hunk 17]
```diff
@@ -487,4 +319,178 @@ public boolean isFinished()
...
+        @Nullable
+        private Block minValues;
+        @Nullable
+        private Block maxValues;
+        @Nullable
+        private BlockPositionComparison minMaxComparison;
+
+        public ChannelFilter(Channel channel, boolean minMaxEnabled, PlanNodeId planNodeId, BlockTypeOperators blockTypeOperators, int channelIndex)
+        {
+            type = channels.get(channelIndex).type;
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:343]: Can we do `type = channel.type` and remove the `channelIndex` argument ?

Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java]: done

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java (hunk 12 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:323]: You can pass a `Callable<Void>` if you don't like `Runnable` :)
non-static inner class is not wrong, just that if it is `static` then it is much easier for someone reading the code to understand what is the linkage between the inner and outer class. Otherwise you need to look closer at inner class implementation to see where and how it accesses the fields and methods of the outer class. This also makes it easier to write unit tests for the inner class if we want to do that at some point.

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java (hunk 10 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java]: My expectation is for the private methods of a class to be grouped together and all the private static things to be after the private things.
Right now what we have is private methods, then private static method, then private static class, then private class, then private method. So some reordering is needed for consistency.

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java (hunk 17 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java]: There is still unused argument `channelIndex`

**Comment by skrzypo987** [comment]
@raunaqmorarka AC
Microbenchmarks - no changes at all
I will run macro benchmarks today

File: core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java [hunk 18]
```diff
@@ -487,4 +322,190 @@ public boolean isFinished()
...
+            if (comparison.compare(block, minValuePosition, currentMin, 0) < 0) {
+                minValues = block.getSingleValueBlock(minValuePosition);
+            }
+            if (comparison.compare(block, maxValuePosition, currentMax, 0) > 0) {
+                maxValues = block.getSingleValueBlock(maxValuePosition);
+            }
+        }
+    }
+
+    private void finishDomainCollectionIfNecessary()
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java:496]: Can we move this method up to where the rest of this class methods are now that we don't reference it directly in ChannelFilter ?

File: core/trino-main/src/test/java/io/trino/operator/TestDynamicFilterSourceOperator.java [hunk 19]
```diff
@@ -133,7 +133,8 @@ private OperatorFactory createOperatorFactory(
         return new DynamicFilterSourceOperator.DynamicFilterSourceOperatorFactory(
                 0,
                 new PlanNodeId("PLAN_NODE_ID"),
-                new DynamicFilterSourceConsumer() {
+                new DynamicFilterSourceConsumer()
```
Comment (raunaqmorarka) [inline core/trino-main/src/test/java/io/trino/operator/TestDynamicFilterSourceOperator.java:136]: looks like unrelated change

**Review by raunaqmorarka — COMMENTED** [review]
lgtm, can land it after we have benchmarking results

File: core/trino-main/src/test/java/io/trino/operator/TestDynamicFilterSourceOperator.java (hunk 19 above)
Comment (skrzypo987) [reply core/trino-main/src/test/java/io/trino/operator/TestDynamicFilterSourceOperator.java:136]: aaaaaaaaagh, not again

**Comment by skrzypo987** [comment]
The benchmarks are in
[Benchmarks DF.pdf](https://github.com/trinodb/trino/files/10352145/Benchmarks.DF.pdf)
and we have a regression. Few queries do regress visibly, mostly tpcds/q93
All of them indeed have joins with multiple channels so this seems to be a correct result.

**Comment by raunaqmorarka** [comment]
> The benchmarks are in
> [Benchmarks DF.pdf](https://github.com/trinodb/trino/files/10352145/Benchmarks.DF.pdf)
> and we have a regression. Few queries do regress visibly, mostly tpcds/q93
> All of them indeed have joins with multiple channels so this seems to be a correct result.

We should benchmark the unpartitioned case as well.

In q93 I suspect earlier the fallback happened quicker due to high cardinality of `sr_ticket_number`, now we probably continue collection of `sr_item_sk` for longer. Could you confirm that is the reason for slowdown ?
In general we're going to try harder to collect dynamic filters after this change, so it's unavoidable that some cases will see a bit more overhead due to DF collection while not ending up with any benefit. There won't be zero regressions guarantee for all workloads even if manage to tweak things to have no regression in the benchmark.

One option is to scale the distinct values limit by some factor depending on the number of columns we start with (e.g. max distinct values * (0.75) ^ (number of DF columns)).
Another option is to look at why DF collection is slow. I suspect there maybe opportunity to improve `TypedSet#add` by making it a batched operations, using `Block#mayHaveNull` etc.

**Review by raunaqmorarka — APPROVED** [review]
